lock_api/src/mutex.rs

```diff
@@ -233,6 +248,20 @@ impl<R: RawMutex, T: ?Sized> Mutex<R, T> {
         }
     }
 
+    /// Attempts to acquire a pinned mutex and returns a pinned guard.
+    /// 
+    /// This function is equivalent to the [`try_lock()`] function, but takes a `Pin<&Mutex>`
+    /// and returns a `Pin<MutexGuard>`. This can be useful if you want to use
+    /// the mutex in a context where the data it protects is pinned, such as a future.
+    #[inline]
+    pub fn try_lock_pinned(self: Pin<&Self>) -> Option<Pin<MutexGuard<'_, R, T>>> {
```

Comment:
This is unsound. `&T` is never pinned. As such if you have `Pin<&Mutex<R, TypeThatNeedsPin>>` then you can simply do `pinned_mutex.try_lock()` to get an `Option<MutexGuard<'_, R, TypeThatNeedsPin>>`, which then allows you to move the value that was supposed to be pinned, thus breaking the pin invariant.

Reply:
This returns a `Pin<MutexGuard<T>>`, not a `MutexGuard<T>`, which is equivalent to a `Pin<&mut T>` and cannot be moved out of soundly.

Reply:
Wait, I see what youre saying. This is unsound. Closing.